src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -206,13 +225,33 @@ private void Stop()
 
                 IPEndPoint localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.LocalAddress);
                 IPEndPoint remoteEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.RemoteAddress);
+                string targetHost = string.Empty;   // compat with SslStream
+                if (connectionInfo.ServerNameLength > 0 && connectionInfo.ServerName != IntPtr.Zero)
+                {
+                    // TBD We should figure out what to do with international names.
+                    targetHost = Marshal.PtrToStringAnsi(connectionInfo.ServerName, connectionInfo.ServerNameLength);
+                    //Console.WriteLine("Listener got connected to {0}", targetHost);
+                }
+
+                SafeMsQuicConfigurationHandle? connectionConfiguration = state.ConnectionConfiguration;
+
+                if (state.AuthenticationOptions.ServerCertificateSelectionCallback != null)
+                {
+                    // ServerCertificateSelectionCallback is synchronous. We will call it as needed when building configuration
+                    connectionConfiguration = SafeMsQuicConfigurationHandle.Create(state.ConnectionOptions, state.AuthenticationOptions, targetHost);
```

Comment:
We probably shouldn't be calling the user's callback on the msquic thread, right?

If we need to do this for now, then that's fine; but we should at least file an issue on this.

Reply:
ok. We already do that for validation. We can change them both but we would end up with quite a bit more complicated code IMHO. Aside from handling exceptions what are the main drawbacks of running from the event handler?